### DISCUSSION TITLE
ZIP scheme yes or no?

### DISCUSSION THREAD

**Discussion by Oldes** [Oldes/Rebol3#54] [Ideas]
In my current Rebol version there is built in support for reading/writing ZIP archives using a dedicated codec. There are several usage examples in the codec's test file:

https://github.com/Oldes/Rebol3/blob/9c128dc365eb941291210ebdc01d7efd2c3fc74e/src/tests/units/codecs-test.r3#L253-L320

My question is, if it makes sense to also create a `zip` scheme? I have a feeling that scheme will not give me so much freedom like the codecs. On the other side, the scheme may make the zip feature more accessible. Is there anybody who wants the scheme and if so, how it should work?

**Comment by rebolek** [comment]
I believe scheme is a better fit for ZIP (and TAR.GZ) than a codec. Both TAR and ZIP are designed as a supersimple filesystem where you can CRUD (Create, Read, Update, Delete) files, so scheme is the way to go IMO.

**Reply by Oldes** [reply]
But how it should look like? With real pseudo usage examples!

**Reply by Oldes** [reply]
The tar codec is not so complete as the zip one...
https://github.com/Oldes/Rebol3/blob/9c128dc365eb941291210ebdc01d7efd2c3fc74e/src/tests/units/codecs-test.r3#L324-L338

**Comment by Oldes** [comment]
There is one big advantage of the schemes, that it is higher level and so one could change internal implementation just by changing the scheme's name... if it would done correctly.

**Comment by luce80** [comment]
Let me try:
```
>> read/as %file.zip 'zip
== [%file-1 %file-2 %file-3]
>> read/part/as %file.zip %file-1 'zip
== "Hello!"
>> remove %file.zip
== [%file-2 %file-3]
>> remove/part %file.zip %file-3
== [%file-2]
>> write/as %file.zip %file-1 'zip
== [%file-2 %file-1]
```
But it could conflict with the codec, and I am not sure it is "correct".

**Reply by Oldes** [reply]
But you are not using any scheme in above code. And such a functionality is already possible with the use of codecs.. like to extract all zip content, you may just do:
```rebol
>> data: load %/Users/oldes/Downloads/imreadmulti_bug.zip 
 [ZIP] Decode ZIP data ( 965 bytes )
 [ZIP] Extracting:  imreadmulti_bug/  bytes: 0 -> 0
 [ZIP] Extracting:  imreadmulti_bug/openmulti.py  bytes: 168 -> 242
 [ZIP] Extracting:  imreadmulti_bug/2page.tiff  bytes: 251 -> 4004
== [
    %imreadmulti_bug/
    [16-Sep-2017/11:44:58 none 0 extra: #{55540500039AF2BC5975780B000104E803000004E8030000} att-ext: 1106051088]
    %imreadmulti_bug/openmulti.py
    [26-Aug-2017/13:13:38 #{
696D706F7274206376320A696D706F7274206E756D7079206173206E700A696D
706F727420707072696E740A0A6D796C697374203D205B5D0A6C6F6164656420
3D206376322E696D726561646D756C7469286D617473203D206D796C6973742C
2066696C656E616D65203D202232706167652E74696666222C20666C61677320
3D206376322E494D524541445F414E59434F4C4F...

```

**Reply by Oldes** [reply]
The use of scheme is more like a virtual filesystem, as @rebolek said... where you don't want to extract every file from a large archive.

It would be more like:
```rebol
archive: open zip://path/to/archive.zip
;; now what to do with the opened archive?
```
There are many actions which are possible on the port. The question is, which to support and how. For example I can imagine to get single record using...
```rebol
data: pick archive %imreadmulti_bug/openmulti.py
;;or
data: read zip://path/to/archive.zip?imreadmulti_bug/openmulti.py
```
But there is still so many questions... like :
* Should there be support for traversing?
* What should be different between `pick` and `select`
* What if we query a directory record?
 etc.

**Comment by luce80** [comment]
>such a functionality is already possible with the use of codecs

Sorry but I was not showing the "functionality" but its different "grammar" in a non-codec approch. I clearly misunderstood your question and didn't understood you were interested in its low-level aspects. That's too deep for me ;) . But I think that if you are seeing it as a (virtual) filesystem you should treat it as such as much as possible. The difficult part is to let the user manipulate the archive also as a "normal" file but if he/she then wants to use the archive functionality he will use your low-level scheme (or the codec if that will remain).
